### ISSUE TITLE
[MRESOLVER-114] ArtifactNotFoundExceptions when building in parallel

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-resolver#890]
**[Rainer Reich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rreich)** opened **[MRESOLVER-114](https://issues.apache.org/jira/browse/MRESOLVER-114?redirect=false)** and commented

We have a multi-module project with quite many modules and many dependencies and observe pretty random `ArtifactNotFoundExceptions` when building in parallel with an empty local repository.

The "funny" thing is that Maven did in fact download the jar that it complained about not finding.

In this example Maven said it could not find `edu.tum.cs:cup-runtime:jar:11a` (see stacktrace below)
But it also said: `Downloaded from central-mirror: [https://mirror.xy.com/repository/maven-all-mirror/edu/tum/cs/cup-runtime/11a/cup-runtime-11a.jar]`.

When looking into the local repository after the failed build we found the cup-runtime.jar in the correct place with the correct checksum.

We tried the following to "fix" the problem:
* build with and without the takari extensions (`takari-local-repository` & `takari-smart-builder`) using `--builder smart`
* also build with `io.takari.aether:aether-connector-okhttp` extension
* only execute goal package instead of install
* build with these properties: `-Daether.connector.basic.threads=1 -Daether.connector.resumeDownloads=false`

But nothing helped.

Unfortunately it seems to be a race condition and we cannot reproduce it consistently but it happens in about 1 out of 5 builds.

```java
...
[2019-11-18T16:46:29.370Z] [INFO] Downloaded from central-mirror: https://mirror.xy.com/repository/maven-all-mirror/edu/tum/cs/cup-runtime/11a/cup-runtime-11a.jar (13 kB at 738 kB/s)
...
[2019-11-18T16:46:30.894Z] [ERROR] Failed to execute goal on project xy: Could not resolve dependencies for project xy: The following artifacts could not be resolved: edu.tum.cs:cup-runtime:jar:11a, org.checkerframework:checker-qual:jar:2.5.7, org.ow2.asm:asm:jar:7.2, cglib:cglib:jar:3.3.0: Could not find artifact edu.tum.cs:cup-runtime:jar:11a -> [Help 1]
[2019-11-18T16:46:30.894Z] org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal on project xy: Could not resolve dependencies for project xy: The following artifacts could not be resolved: edu.tum.cs:cup-runtime:jar:11a, org.checkerframework:checker-qual:jar:2.5.7, org.ow2.asm:asm:jar:7.2, cglib:cglib:jar:3.3.0: Could not find artifact edu.tum.cs:cup-runtime:jar:11a
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.getDependencies (LifecycleDependencyResolver.java:269)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.resolveProjectDependencies (LifecycleDependencyResolver.java:147)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.MojoExecutor.ensureDependenciesAreResolved (MojoExecutor.java:248)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:202)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
[2019-11-18T16:46:30.894Z]     at io.takari.maven.builder.smart.SmartBuilderImpl.buildProject (SmartBuilderImpl.java:205)
[2019-11-18T16:46:30.894Z]     at io.takari.maven.builder.smart.SmartBuilderImpl$ProjectBuildTask.run (SmartBuilderImpl.java:77)
[2019-11-18T16:46:30.894Z]     at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:515)
[2019-11-18T16:46:30.894Z]     at java.util.concurrent.FutureTask.run (FutureTask.java:264)
[2019-11-18T16:46:30.894Z]     at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1128)
[2019-11-18T16:46:30.894Z]     at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:628)
[2019-11-18T16:46:30.894Z]     at java.lang.Thread.run (Thread.java:834)
[2019-11-18T16:46:30.894Z] Caused by: org.apache.maven.project.DependencyResolutionException: Could not resolve dependencies for project xy: The following artifacts could not be resolved: edu.tum.cs:cup-runtime:jar:11a, org.checkerframework:checker-qual:jar:2.5.7, org.ow2.asm:asm:jar:7.2, cglib:cglib:jar:3.3.0: Could not find artifact edu.tum.cs:cup-runtime:jar:11a
[2019-11-18T16:46:30.894Z]     at org.apache.maven.project.DefaultProjectDependenciesResolver.resolve (DefaultProjectDependenciesResolver.java:209)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.getDependencies (LifecycleDependencyResolver.java:243)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.resolveProjectDependencies (LifecycleDependencyResolver.java:147)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.MojoExecutor.ensureDependenciesAreResolved (MojoExecutor.java:248)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:202)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
[2019-11-18T16:46:30.894Z]     at io.takari.maven.builder.smart.SmartBuilderImpl.buildProject (SmartBuilderImpl.java:205)
[2019-11-18T16:46:30.894Z]     at io.takari.maven.builder.smart.SmartBuilderImpl$ProjectBuildTask.run (SmartBuilderImpl.java:77)
[2019-11-18T16:46:30.894Z]     at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:515)
[2019-11-18T16:46:30.894Z]     at java.util.concurrent.FutureTask.run (FutureTask.java:264)
[2019-11-18T16:46:30.894Z]     at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1128)
[2019-11-18T16:46:30.894Z]     at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:628)
[2019-11-18T16:46:30.894Z]     at java.lang.Thread.run (Thread.java:834)
[2019-11-18T16:46:30.894Z] Caused by: org.eclipse.aether.resolution.DependencyResolutionException: The following artifacts could not be resolved: edu.tum.cs:cup-runtime:jar:11a, org.checkerframework:checker-qual:jar:2.5.7, org.ow2.asm:asm:jar:7.2, cglib:cglib:jar:3.3.0: Could not find artifact edu.tum.cs:cup-runtime:jar:11a
[2019-11-18T16:46:30.894Z]     at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies (DefaultRepositorySystem.java:357)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.project.DefaultProjectDependenciesResolver.resolve (DefaultProjectDependenciesResolver.java:202)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.getDependencies (LifecycleDependencyResolver.java:243)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.resolveProjectDependencies (LifecycleDependencyResolver.java:147)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.MojoExecutor.ensureDependenciesAreResolved (MojoExecutor.java:248)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:202)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
[2019-11-18T16:46:30.894Z]     at io.takari.maven.builder.smart.SmartBuilderImpl.buildProject (SmartBuilderImpl.java:205)
[2019-11-18T16:46:30.894Z]     at io.takari.maven.builder.smart.SmartBuilderImpl$ProjectBuildTask.run (SmartBuilderImpl.java:77)
[2019-11-18T16:46:30.894Z]     at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:515)
[2019-11-18T16:46:30.894Z]     at java.util.concurrent.FutureTask.run (FutureTask.java:264)
[2019-11-18T16:46:30.894Z]     at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1128)
[2019-11-18T16:46:30.894Z]     at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:628)
[2019-11-18T16:46:30.894Z]     at java.lang.Thread.run (Thread.java:834)
[2019-11-18T16:46:30.894Z] Caused by: org.eclipse.aether.resolution.ArtifactResolutionException: The following artifacts could not be resolved: edu.tum.cs:cup-runtime:jar:11a, org.checkerframework:checker-qual:jar:2.5.7, org.ow2.asm:asm:jar:7.2, cglib:cglib:jar:3.3.0: Could not find artifact edu.tum.cs:cup-runtime:jar:11a
[2019-11-18T16:46:30.894Z]     at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve (DefaultArtifactResolver.java:424)
[2019-11-18T16:46:30.894Z]     at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts (DefaultArtifactResolver.java:229)
[2019-11-18T16:46:30.894Z]     at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies (DefaultRepositorySystem.java:340)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.project.DefaultProjectDependenciesResolver.resolve (DefaultProjectDependenciesResolver.java:202)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.getDependencies (LifecycleDependencyResolver.java:243)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.resolveProjectDependencies (LifecycleDependencyResolver.java:147)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.MojoExecutor.ensureDependenciesAreResolved (MojoExecutor.java:248)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:202)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
[2019-11-18T16:46:30.894Z]     at io.takari.maven.builder.smart.SmartBuilderImpl.buildProject (SmartBuilderImpl.java:205)
[2019-11-18T16:46:30.894Z]     at io.takari.maven.builder.smart.SmartBuilderImpl$ProjectBuildTask.run (SmartBuilderImpl.java:77)
[2019-11-18T16:46:30.894Z]     at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:515)
[2019-11-18T16:46:30.894Z]     at java.util.concurrent.FutureTask.run (FutureTask.java:264)
[2019-11-18T16:46:30.894Z]     at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1128)
[2019-11-18T16:46:30.894Z]     at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:628)
[2019-11-18T16:46:30.894Z]     at java.lang.Thread.run (Thread.java:834)
[2019-11-18T16:46:30.894Z] Caused by: org.eclipse.aether.transfer.ArtifactNotFoundException: Could not find artifact edu.tum.cs:cup-runtime:jar:11a
[2019-11-18T16:46:30.894Z]     at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve (DefaultArtifactResolver.java:414)
[2019-11-18T16:46:30.894Z]     at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts (DefaultArtifactResolver.java:229)
[2019-11-18T16:46:30.894Z]     at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveDependencies (DefaultRepositorySystem.java:340)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.project.DefaultProjectDependenciesResolver.resolve (DefaultProjectDependenciesResolver.java:202)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.getDependencies (LifecycleDependencyResolver.java:243)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.LifecycleDependencyResolver.resolveProjectDependencies (LifecycleDependencyResolver.java:147)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.MojoExecutor.ensureDependenciesAreResolved (MojoExecutor.java:248)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:202)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
[2019-11-18T16:46:30.894Z]     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
[2019-11-18T16:46:30.894Z]     at io.takari.maven.builder.smart.SmartBuilderImpl.buildProject (SmartBuilderImpl.java:205)
[2019-11-18T16:46:30.894Z]     at io.takari.maven.builder.smart.SmartBuilderImpl$ProjectBuildTask.run (SmartBuilderImpl.java:77)
[2019-11-18T16:46:30.894Z]     at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:515)
[2019-11-18T16:46:30.894Z]     at java.util.concurrent.FutureTask.run (FutureTask.java:264)
[2019-11-18T16:46:30.894Z]     at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1128)
[2019-11-18T16:46:30.894Z]     at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:628)
[2019-11-18T16:46:30.894Z]     at java.lang.Thread.run (Thread.java:834)
```

---

**Affects:** 1.4.2

**Attachments:**
- [BasicRepositoryConnector.java](https://issues.apache.org/jira/secure/attachment/13005267/BasicRepositoryConnector.java) (_23.54 kB_)
- [DefaultArtifactResolver.java](https://issues.apache.org/jira/secure/attachment/13005266/DefaultArtifactResolver.java) (_34.04 kB_)
- [maven-debug-log.txt](https://issues.apache.org/jira/secure/attachment/12986349/maven-debug-log.txt) (_3.16 MB_)
- [mvn-build-120.log](https://issues.apache.org/jira/secure/attachment/13005264/mvn-build-120.log) (_49.51 MB_)
- [mvn-build-120-sorted.txt](https://issues.apache.org/jira/secure/attachment/13005268/mvn-build-120-sorted.txt) (_19.42 kB_)
- [mvn-build-134.log](https://issues.apache.org/jira/secure/attachment/13005308/mvn-build-134.log) (_43.54 MB_)
- [mvn-build-134-sorted.txt](https://issues.apache.org/jira/secure/attachment/13005306/mvn-build-134-sorted.txt) (_30.98 kB_)
- [mvn-build-144.log](https://issues.apache.org/jira/secure/attachment/13005311/mvn-build-144.log) (_38.67 MB_)
- [mvn-build-144-sorted.txt](https://issues.apache.org/jira/secure/attachment/13005312/mvn-build-144-sorted.txt) (_25.77 kB_)

**Issue Links:**
- [MRESOLVER-123](https://issues.apache.org/jira/browse/MRESOLVER-123) Provide a global locking sync context by default
 (_**"is fixed by"**_)
- [MRESOLVER-131](https://issues.apache.org/jira/browse/MRESOLVER-131) Introduce a Redisson-based SyncContextFactory
 (_**"is fixed by"**_)
- [MNG-5315](https://issues.apache.org/jira/browse/MNG-5315) Artifact resolution sporadically fails in parallel builds
 (_**"causes"**_)

**Remote Links:**
- [GitHub Pull Request #41
](https://github.com/apache/maven-resolver/pull/41)
- [GitHub Pull Request #42
](https://github.com/apache/maven-resolver/pull/42)
- [GitHub Pull Request #60
](https://github.com/apache/maven-resolver/pull/60)
- [[v0.11.2] LockingSyncContext not used
](https://github.com/takari/takari-local-repository/issues/11)
- [[v0.11.2] DefaultFileManager thread safe?
](https://github.com/takari/takari-local-repository/issues/12)

**Backported to:** [1.6.0](https://github.com/apache/maven-resolver/milestone/47?closed=1), [1.5.0](https://github.com/apache/maven-resolver/milestone/46?closed=1)

3 votes, 10 watchers

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

This could be a dup of those cases I have already commented on. Please search for similiar issue. A solution will require a significant amount of time and concurrent expertise. At least the former is not available.

It would really help if you could provide a debug output with `-X` for that failure.

**Comment by jira-importer** [comment]
**[Rainer Reich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rreich)** commented

It looks like MNG-5315 which has already been closed. I couldn't find any other issue describing this problem.

I only found some mailing list archive with the same problem from 2015: http://mail-archives.apache.org/mod_mbox/maven-users/201504.mbox/%3c1287435039.496078.1428690919489.JavaMail.yahoo@mail.yahoo.com%3e

**Comment by jira-importer** [comment]
**[Rainer Reich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rreich)** commented

I added a  [^maven-debug-log.txt]  where the artifact `com.acme.cms:acme-test-logging-config` could not be found. This is not a public artifact but a SNAPSHOT artifact coming from an internal nexus, but the error looks exactly the same. Hope that helps.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

I have a question based on what you have written:

**The "funny" thing is that Maven did in fact download the jar that it complained about not finding**

Does you build correctly work if you first clean your local repository and make a simple `mvn clean package` ?

Furthermore you are already using takari local repository handler based on the log ?

**Comment by jira-importer** [comment]
**[Rainer Reich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rreich)** commented

We always build with an empty local repository and use the `takari-local-repository` and `takari-smart-builder` extensions by default.
The issue only occurs when building in parallel, but for the goal `package` as well as for `install`.
So yes, `mvn clean package` works, but `mvn clean package -T2` shows these problems, too.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Can you please make a check without takari parts ?

**Comment by jira-importer** [comment]
**[Rainer Reich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rreich)** commented

We already tried that (see first bullet point of our "fix" approaches) and the same problems occurred.

**Comment by jira-importer** [comment]
**[Rainer Reich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rreich)** commented

We think we might have found the problem which is inside the `org.eclipse.aether.internal.impl.DefaultArtifactResolver#resolve()` method:

This method first looks into the local repository calling `lrm.find(...)` and then downloads the artifacts that are still unresolved in `performDownloads()`.
If after this, the artifacts stored in the results don't have a file set an `ArtifactNotFoundException` is thrown.
(See https://github.com/apache/maven-resolver/blob/master/maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultArtifactResolver.java#L409)

Now let's say we have two threads A and B that are both entering the `resolve()` method and want to resolve the same artifact.
They could be executed as follows:
1. A is looking into its local repository and doesn't find the artifact;
2. B is running and also doesn't find it in its local repository;
3. B downloads it, finishes the download and marks the item as resolved;
4. A is continued and doesn't download anything because the item is already resolved;
5. Then A's result will never have an artifact file set and A runs into the exception.

The simplest fix is to synchronize the whole `resolve` method, so that A won't get interrupted.
Theoretically not optimal, but dead-simple and we could not measure any significant performance impact.
(See https://github.com/apache/maven-resolver/pull/42/files)

Another fix (workaround?) we tried, was to look up the artifact in the local repository again after `performDownloads()`.
(See https://github.com/apache/maven-resolver/pull/41/files)

Both fixes seem to be working. While we managed to set up our workspace so that the original method produced the ArtifactNotFoundException in nearly 50% of the runs,
the fixed versions never showed any issues.

Probably you will come up with a more sophisticated approach for fixing this problem, but we hope we could help to shed some light on this.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Rainer Reich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rreich), you analysis sounds reasonable. Can you tell which of these is faulty: https://github.com/apache/maven-resolver/blob/8d2c3752af1b724ea224bd7006ca9ab911db3a87/maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultArtifactResolver.java#L412-L416 or https://github.com/apache/maven-resolver/blob/8d2c3752af1b724ea224bd7006ca9ab911db3a87/maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultArtifactResolver.java#L263-L267.

I don't fully understand the transition fron 4 to 5. If the nothing has to be downloaded and the item is available via `LOCAL_PATH` how can it fail then? Why does the artifact not contain a file?

**Comment by jira-importer** [comment]
**[Rainer Reich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rreich)** commented

Thanks for taking a look and for moving the ticket!
Regarding the transition from 4. to 5.:
4. does not mean that this artifact has a localPath or no file. It is a regular artifact, but its files have already been downloaded by Thread B in the meantime and A simply does not know about this.
A doesn't see the artifact or its files which have been downloaded by B so that https://github.com/apache/maven-resolver/blob/8d2c3752af1b724ea224bd7006ca9ab911db3a87/maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultArtifactResolver.java#L409 will evaluate to false.
This is where my PullRequest #41 comes into play, which checks again right after L409, if the artfacts are available in the local repository now, which they are because B just put them there.

BTW: We are using the solution from #41 in our CI for quite some time now and have never seen this bug again, so it really seems to work.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

That's still a bit strange for me. If B does not see what it does, why not simply overwrite (redownload) and done?

**Comment by jira-importer** [comment]
**[Rainer Reich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rreich)** commented

That is basically what we are doing, but instead of re-downloading, it is sufficient to just look into the local repository again.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I am still going through the issue and your change. I am having the feeling (I might be wrong) that your sophisticated patch just fixes the symptom and not the cause. Based on your analysis I understand that the issue is here:

```java
if ( isLocallyInstalled( local, versionResult ) )
{
    if ( local.getRepository() != null )
    {
        result.setRepository( local.getRepository() );
    }
    else
    {
        result.setRepository( lrm.getRepository() );
    }
    try
    {
        artifact = artifact.setFile( getFile( session, artifact, local.getFile() ) );
        result.setArtifact( artifact );
        artifactResolved( session, trace, artifact, result.getRepository(), null );
    }
    catch ( ArtifactTransferException e )
    {
        result.addException( e );
    }
    if ( !local.isAvailable() )
    {
        /*
         * NOTE: Interop with simple local repository: An artifact installed by a simple local repo manager
         * will not show up in the repository tracking file of the enhanced local repository. If however the
         * maven-metadata-local.xml tells us the artifact was installed locally, we sync the repository
         * tracking file.
         */
        lrm.add( session, new LocalArtifactRegistration( artifact ) );
    }
    continue;
}
else if ( local.getFile() != null )
{
    LOGGER.debug( "Verifying availability of {} from {}", local.getFile(), repos );
}
```

So what I understand is that thread A has passed `performDownloads()` as well the forthcoming loop over `results`.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I want to release version 1.4.3 somehwere next week and I'd like include a fix for this. Please let me know how we can proceed with this.

**Comment by jira-importer** [comment]
**[Rainer Reich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rreich)** commented

It might be a symptom fix. I'm trying to understand it again, but by reading code I'm unable to see how Thread A is affected by Thread B at all (but obviously it is).

I couldn't reproduce the same exception again, but when running with the original Maven I got a ~50% chance of failure in our CI with some kind of ArtifactNotFoundException.
Unfortunately I wasn't able to reproduce it often enough on my machine to debug, so I can only say, the resolve method has race conditions that we cannot observe after applying one of the two provided patches anymore.

So, we would be very happy if you would merge the PR #42 anyway. It really helps us and it shouldn't do any harm.
If you understand where the concurrency and shared state is hiding, please give me a pointer.
I'm giving up for today. Happy Weekend!

**Comment by jira-importer** [comment]
**[Eva Müller](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=emueller-coremedia)** commented

[Rainer Reich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rreich) I guess, you want to have [PR #41](https://github.com/apache/maven-resolver/pull/41) to be merged, won't you?!

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

> It might be a symptom fix. I'm trying to understand it again, but by reading code I'm unable to see how Thread A is affected by Thread B at all (but obviously it is).

Same here.

> Unfortunately I wasn't able to reproduce it often enough on my machine to debug, so I can only say, the resolve method has race conditions that we cannot observe after applying one of the two provided patches anymore.

I assume that you can't provide a stripped down project to reproduce issue on my machine?

> If you understand where the concurrency and shared state is hiding, please give me a pointer.

Unfortunately, I haven't yet found a possible spot for race conditions. I am inclined to merge PR #41 for now since we don't understand the cause.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I have put some effort into it analyzed the issue and it is disappointing. First of all, I think we suffer from [this](https://stackoverflow.com/q/21739080/696632) problem and both PRs are wrong while the retry approach is, IMHO, even worse even if it/both "solves" the issue. I was not able to find logically the code spot in question, but there **is** a race condition. Unfortunately, Resolver has very little logging which would help us to narrow down the issue.

Now let me go into detail what I have figured out:
When we go the callstack up from #resolve() we land in public ... resolveArtifacts(...) which does the following:

```java
try ( SyncContext syncContext = syncContextFactory.newInstance( session, false ) )
{
	Collection<Artifact> artifacts = new ArrayList<>( requests.size() );
	for ( ArtifactRequest request : requests )
	{
		if ( request.getArtifact().getProperty( ArtifactProperties.LOCAL_PATH, null ) != null )
		{
			continue;
		}
		artifacts.add( request.getArtifact() );
	}

	syncContext.acquire( artifacts, null );

	return resolve( session, requests );
}
```

Wait a minute, there is synchronization already, why does it still fail? Well...let's have a look at SyncContextFactory and SyncContext first:
SyncContextFactory shall create a reentrant read write lock for the given session. The boolean flag says read or write. SyncContext#acquire() shall perform the lock operation on the values (GAV keys) (!) of the given artifacts/metadata. This approach is used everywhere in the code where interaction with the filesytem is necessary. So #resolve() is **not** intended to be thread safe and it would be wrong to try to change it. Back to the question: When we now look for all impls of SyncContext we find only DefaultSyncContext which does nothing, i.e., no synchronization at all. Why do other operations not fail with SyncContext? I assume that most either don't use a degree a concurrency where it would fail on the same artifacts or it runs so fast that we don't see race conditions.
How can this now be fixed? We have two challenges here:
1. #newInstance() implies that syncContext is always new (other identity), but how shall #acquire() work multithreaded if no global locking on the artifacts can be maintained? Weird. The only solution which comes to my mind is to initialize a suitable lock implementation when #newInstance() is called in a synchronized block and put into RepositorySystemSession#getData().If this cannot be maintained we still will have the same race condition.
2. A suitable lock approach, if selfwritted, would require something like a concurrent hash map with weak refs onto muteces which represent the value (hashcode) of an artifact. On #acquire() the lock for that mutex is obtained and #lock() called. When the try block is left #release() is called automatically.
   Alternatively, one needs to use a KeyedReadWriteLock or StripedReadWriteLock because we cannot maintain a lock per artifacts. This could end with thousands of locks.

I think even for someone who does concurrency on a daily basis, it would take some days to solve these points, for me whose concurrency skills are weak, weeks. I cannot provide this one.
It has also to be noted that DefaultArtifactResolver is not marked with `@Singleton`. I don't know whether there could be multiple instances used in Maven which means that only a global lock will work because synchronized will not have any effect or if SyncContextFactory is a singleton it could hold the global lock and pass it to the SyncContext objects.

See also: https://guava.dev/releases/19.0/api/docs/com/google/common/util/concurrent/Striped.html

A completely alternative approach is to write lock files for those objects instead of an in-memory lock. SyncContext says "Note that the level of actual synchronization is subject to the implementation and might range from OS-wide to none."

Unless someone can provide something very smart, I will do the following:

* Add FIXMEs to all invocations of SyncContextFactory#newInstance() to a new issue which describes this problem.
* Accept the compromise of PR #41.

Opinions?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte), can you share your opinion please?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

According to `AetherModule` `SyncContextFactory` is a singleton. Even if we'd have multiple DefaultArtifactResolver instances this would be OK.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

This basically means that the their `SyncContext` impl does not work. At first glance, their implementation looks exactly how I expect the file locking to look like. This means that this issue is invalid because of the former. We need to reach out to Takari or adopt this code into ours and try to figure out why the locking does not work as expected.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Eva Müller](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=emueller-coremedia), did you actually verify that the lock files are created at all?

**Comment by jira-importer** [comment]
**[Eva Müller](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=emueller-coremedia)** commented

Here are the next log files. All with http DEBUG and without builder smart...

### Build 134 (ArtifactNotFoundException)

* `mvn clean ... deploy -T3 ...` => **without using smart builder**, i.e. takari-local-manager should not be used
* [^mvn-build-134.log]
* [^mvn-build-134-sorted.txt]
* 

```
Caused by: org.eclipse.aether.transfer.ArtifactNotFoundException: Failure to find org.hibernate:hibernate-validator:jar:5.4.3.Final in https://nexus.global-jenkins.vm/repository/local-releases was cached in the local repository, resolution will not be reattempted until the update interval of global-releases has elapsed or updates are forced
```

### Build 144 (java.io.FileNotFoundException)

* `mvn clean ... deploy -T5 ...` => **without using smart builder**, i.e. takari-local-manager should not be used
* [^mvn-build-144.log]
* [^mvn-build-144-sorted.txt]
* 

```
Caused by: java.io.FileNotFoundException: /my-workspace/workspace/cms/apps/studio-server/.m2/repository/com/typesafe/netty/netty-reactive-streams/2.0.3/netty-reactive-streams-2.0.3.pom.part (No such file or directory)
```

### Built with/by

* [^DefaultArtifactResolver.java] with additional log statements
* [^BasicRepositoryConnector.java] with additional log statements
* `MAVEN_OPTS` set to `"-Dorg.apache.commons.logging.Log=org.apache.commons.logging.impl.SimpleLog -Dorg.slf4j.simpleLogger.log.org.apache.maven.wagon.providers.http.httpclient=DEBUG"` (Hope these are the correct HttpClient debug log statements for headers)

**Comment by jira-importer** [comment]
**[Eva Müller](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=emueller-coremedia)** commented

Until now, I built our workspace 14 times

with the **master** `takari-local-repository` Maven extension [0.11.4-SNAPSHOT](https://github.com/takari/takari-local-repository/blob/master/pom.xml#L18) with additional log statements

without a single failure. That's annoying ;-) .

---

FYI: Building our workspace w/o `--builder smart` failed 3 times of 14 builds.

**Comment by jira-importer** [comment]
**[Rainer Reich](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rreich)** commented

I took another look at the Takari synchronization and this pattern and the SyncContext itself are not usable for synchronization of multiple threads as there are multiple instances of the Resolver and the SyncContext. The code is just written in a way that seems to intentionally wanting to prevent thread-safety...
Therefore I created a new PullRequest (https://github.com/apache/maven-resolver/pull/60) that makes the lock a static member of the DefaultArtifactResolver and synchronizes the whole resolve-Method. This has to work, right? And I could not see any performance impact.
If you fear performance impact, this could be an optional feature using some sort of feature flag to enable synchronization.
What do you think?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Please test a new, global locking approach with MRESOLVER-123. It is suited for concurrent access in **one** VM.

**Comment by jira-importer** [comment]
**[Philippe GRANET](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=phgr100x)** commented

I found these two maven extensions that try to fix concurrency problem:
* https://github.com/evanchsa/maven-parallel-resolver

It works in one JVM, it synchronizes the whole resolve-Method, but it has performance impact since all downloads are sequential, it disable maven parallel download (mvn -T ... and also mvn -Dmaven.artifact.threads=...)
* https://github.com/murkaje/robust-maven-cache

Not tested

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Philippe GRANET](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=phgr100x), the first one is logically broken because does not cover all cases. It synchs the wrong spot. The second one looks more sophisticated, but still does not use the sync context at all. I don't see neither as real anternatives. Please try the MRESOLVER-123 branch if you suffer from this.
